### PR TITLE
[MM-48973] Enable test that was fixed separately

### PR DESCRIPTION
**Pull request by nickmisasi**
#### Summary
There was a race condition around the time this test was reported as flaky. This race condition has since been fixed in https://github.com/mattermost/enterprise/pull/1324

I'm not 100% sure this actually caused the issue but I've been unable to reproduce it locally.

To be safe, I've wrapped this call in a Cloud check to ensure it doesn't run in non-cloud contexts. 

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-48973

#### Release Note

```release-note
None
```

### CODE REVIEW

**Review by agarciamontoro — APPROVED** [review]
> I'm not 100% sure this actually caused the issue but I've been unable to reproduce it locally.

Let's see if it does solve it. If I see it failing again, I'll ping you :) Thanks for the fix!

**Review by agnivade — CHANGES_REQUESTED** [review]
Sorry, but this is not the right fix. We need to be more diligent and thorough here.

> I'm not 100% sure this actually caused the issue but I've been unable to reproduce it locally.

https://github.com/mattermost/enterprise/pull/1324 is not a race condition. In this context, a race condition refers to a data race where the same memory location is being read and written from different threads.

How are you trying to reproduce this? Are you simply running the test normally? That won't detect race conditions. You need to use the `-race` flag. I just ran it and it failed immediately. Please reach out to us if you are failing to reproduce something, and we would be happy to help.

The test clearly shows the line numbers affected.

We can see that we are passing `ruser.Id` to `th.App.AddUserToTeam`, and using the same `ruser` in `a.SendSubscriptionHistoryEvent(ruser.Id)` which is inside a goroutine. This is what causes the race.

The fix should be straightforward: just remove the goroutine. In our style guide, we have [mentioned](https://developers.mattermost.com/contribute/more-info/server/style-guide/#default-to-sync-instead-of-async)
> Do not create one-off goroutines without knowing when/how they exit.

What was the reason behind executing this in a goroutine? If the API call doesn't take very long, then I'd just say to remove the goroutine. Makes the code simpler and fixes the race condition. Guarding it with `a.Channels().License().IsCloud()` is fine as well.

If however, you feel that the usage of goroutine is mandatory here, then we need to pass a copy of the userID to the goroutine like so:
```diff
+++ b/app/user.go
@@ -321,12 +321,12 @@ func (a *App) createUserOrGuest(c request.CTX, user *model.User, guest bool) (*m
        // table in CWS. This is then used to calculate how much the customers have to pay in addition for the extra users. If the
        // workspace is currently on a monthly plan, then this function will not do anything.

-       go func() {
-               _, err := a.SendSubscriptionHistoryEvent(ruser.Id)
+       go func(userID string) {
+               _, err := a.SendSubscriptionHistoryEvent(userID)
                if err != nil {
                        c.Logger().Error("Failed to create/update the SubscriptionHistoryEvent", mlog.Err(err))
                }
-       }()
+       }(ruser.Id)
```

In general, always try to _not_ use a goroutine unless you understand all the implications and how/when does the goroutine exit. As you can see, it causes more issues than what it solves.

**Comment by agnivade** [comment]
> the go routine is necessary for UX in the event of retries - we don't want timeouts or long wait times for something that is happening in the background and really of no concern to the user.

Understood, in that case you really need a separate long-running goroutine with a buffered channel that can be started and stopped gracefully. Right now, there is no limit to the number of goroutines that can be spawned (which can cause perf issues) and there is no control over how they are closed (when you shut down the server, there is no guarantee that they will be completed).

These are the problems with one-off goroutines.

File: app/user.go [hunk 1]
```diff
@@ -321,12 +321,14 @@ func (a *App) createUserOrGuest(c request.CTX, user *model.User, guest bool) (*m
...
 	// workspace is currently on a monthly plan, then this function will not do anything.
 
-	go func() {
-		_, err := a.SendSubscriptionHistoryEvent(ruser.Id)
-		if err != nil {
-			c.Logger().Error("Failed to create/update the SubscriptionHistoryEvent", mlog.Err(err))
-		}
-	}()
+	if a.Channels().License().IsCloud() {
+		go func(userId string) {
```
Comment (agnivade) [inline app/user.go:325]: Nit: we prefer `userID` over `userId`: https://developers.mattermost.com/contribute/more-info/server/style-guide/#hahahugoshortcode-s8-hbhb

**Review by agnivade — APPROVED** [review]

**Comment by agnivade** [comment]
`platform.Go` is actually not a good pattern since that creates unbounded goroutines. From what I understand, the best suited would be a single running goroutine using a buffered channel that listens on requests and executes them. Along with a proper shutdown mechanism that is connected to server shutdown.

**Comment by nickmisasi** [comment]
> Understood, in that case you really need a separate long-running goroutine with a buffered channel that can be started and stopped gracefully. Right now, there is no limit to the number of goroutines that can be spawned (which can cause perf issues) and there is no control over how they are closed (when you shut down the server, there is no guarantee that they will be completed).

Makes sense! As I mentioned earlier, I've got some changes coming down the pipeline to address this whole flow and I expect we'll be able to get rid of the go routine in the process. I'm going to leave it as is in this PR for now and take another look when I get to it in the future

**Pull request merged**